README
CPU-G

CPU-G is a Python 3 and GTK+3 application that shows some useful information about your hardware. It collects and displays information about your CPU, RAM, Motherboard, some general information about your system and more.

Copyright © 2009  Fotis Tsamis <ftsamis@gmail.com>.
Copyright © 2016-2019  Lorenzo Carbonell (aka atareao) <lorenzo.carbonell.cerezo@gmail.com>

Site: http://cpug.sourceforge.net/


debian/changelog
cpu-g (0.16.3-1ubuntu1) UNRELEASED; urgency=medium

  * TypeError: cannot use a string pattern on a bytes-like object # 10
  * arch linux problems (python3.8) # 12 by @s0me0ther

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 18 Apr 2020 21:37:32 +0200

cpu-g (0.16.3-1ubuntu0) focal; urgency=medium

  * Fixed for Debian 9 and MATE #4 by @rstephan
  * Added catalan translation #6 by @marcelcosta
  * Added basque translation #7 by @xabiliza
  * Fixed minor bugs

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 18 Apr 2020 21:33:42 +0200

cpu-g (0.16.1-1ubuntu0) cosmic; urgency=medium

  * Added spanish translation

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 03 Feb 2019 10:46:50 +0100

cpu-g (0.15.2-1ubuntu0) cosmic; urgency=medium

  * fixed bugs #2, #3

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Fri, 01 Feb 2019 20:43:22 +0100

cpu-g (0.15.1-1ubuntu0) zesty; urgency=medium

  * fixed bug zero division #1 by @tonyfarney 

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Tue, 26 Sep 2017 13:32:21 +0200

cpu-g (0.15.0-1ubuntu0) xenial; urgency=medium

  * Fixed a bug when no battery in computer
  * Added more info to battery
  * Added locale to number

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 23 Oct 2016 11:34:47 +0200

cpu-g (0.14.0-1ubuntu0) xenial; urgency=medium

  * Fixed a bug with Xorg version

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 16 Oct 2016 19:51:21 +0200

cpu-g (0.13.0-1ubuntu0) xenial; urgency=medium

  * Added Disks info

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 01 Oct 2016 08:38:41 +0200

cpu-g (0.12.0-1ubuntu0) xenial; urgency=medium

  * Added Swap info
  * Changed battery info to use dbus history

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 10 Sep 2016 08:20:27 +0200

cpu-g (0.11.2-1ubuntu0) xenial; urgency=medium

  * Fixed nvidia logo

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 08 Sep 2016 20:46:34 +0200

cpu-g (0.11.1-1ubuntu3) xenial; urgency=medium

  * Added wmctrl dependecy

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 08 Sep 2016 07:48:00 +0200

cpu-g (0.11.1-1ubuntu2) xenial; urgency=medium

  * New features,
  * - System:
  *   - Added resolution info
  *   - Added distribution info
  *   - Added Desktop environment info
  *   - Added window manager info
  *   - Added distribution logo
  * - Battery
  *   - Monitor battery
  *   - Added manufacturer info
  *   - Added model info
  *   - Added serial number info
  *   - Added tecnology info
  *   - Added estimated time remaining
  *   - Added estimated battery discharge time
  *   - Added battery level
  *   - Added graphic to show the evolution of dicharge


 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 04 Sep 2016 20:37:04 +0200

cpu-g (0.10.0-1ubuntu0) wily; urgency=medium

  * Added graphic card info

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 02 Mar 2016 23:20:00 +0100

cpu-g (0.9.2-1ubuntu0) wily; urgency=medium

  * First release

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 02 Mar 2016 07:52:47 +0100


src/test.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
#
# CPU-G is a program that displays information about your CPU,
# RAM, Motherboard and some general information about your System.
#
# Copyright © 2009  Fotis Tsamis <ftsamis at gmail dot com>.
# Copyright © 2016-2019  Lorenzo Carbonell (aka atareao)
# <lorenzo.carbonell.cerezo at gmail dot com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from gi.repository import Gtk

from matplotlib.figure import Figure
from matplotlib.backends.backend_gtk3agg import FigureCanvasGTK3Agg as
FigureCanvas

win = Gtk.Window()
win.connect("delete-event", Gtk.main_quit)
win.set_default_size(800, 600)
win.set_title("Embedding in GTK")

f = Figure(figsize=(5, 4), dpi=100)
a = f.add_subplot(111)
t = [1, 2, 3, 4]
s = [1, 4, 9, 16]
a.plot(t, s)

sw = Gtk.ScrolledWindow()
win.add(sw)
# A scrolled window border goes outside the scrollbars and viewport
sw.set_border_width(10)

canvas = FigureCanvas(f)  # a Gtk.DrawingArea
canvas.set_size_request(400, 300)
sw.add_with_viewport(canvas)

win.show_all()
Gtk.main()


src/comun.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
#
# CPU-G is a program that displays information about your CPU,
# RAM, Motherboard and some general information about your System.
#
# Copyright © 2009  Fotis Tsamis <ftsamis at gmail dot com>.
# Copyright © 2016-2019  Lorenzo Carbonell (aka atareao)
# <lorenzo.carbonell.cerezo at gmail dot com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import sys
import locale
import gettext


def is_package():
    return __file__.find('src') < 0

######################################


APP = 'cpu-g'
APPNAME = 'CPU-G'

# check if running from source
if is_package():
    ROOTDIR = '/opt/extras.ubuntu.com/cpu-g/share'
    LANGDIR = os.path.join(ROOTDIR, 'locale-langpack')
    APPDIR = os.path.join(ROOTDIR, APP)
    CHANGELOG = os.path.join(APPDIR, 'changelog')
    ICONDIR = os.path.join(ROOTDIR, 'icons')
    ICON = os.path.join(ICONDIR, 'cpu-g.png')
    LOGOSDIR = os.path.join(APPDIR, 'logos')
    DISTROSDIR = os.path.join(APPDIR, 'distros')
    GRAPHICCARDDIR = os.path.join(APPDIR, 'graphic_card')
    BATTERY_MONITOR = os.path.join(ROOTDIR, 'monitor_battery.py')
else:
    ROOTDIR = os.path.dirname(__file__)
    LANGDIR = os.path.normpath(os.path.join(ROOTDIR, '../template1'))
    APPDIR = ROOTDIR
    DEBIANDIR = os.path.normpath(os.path.join(ROOTDIR, '../debian'))
    CHANGELOG = os.path.join(DEBIANDIR, 'changelog')
    ICON = os.path.normpath(os.path.join(ROOTDIR, '../data/icons/cpu-g.png'))
    LOGOSDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/logos'))
    DISTROSDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/distros'))
    GRAPHICCARDDIR = os.path.normpath(os.path.join(ROOTDIR,
                                                   '../data/graphic_card'))
    BATTERY_MONITOR = os.path.join(ROOTDIR, 'monitor_battery.py')

f = open(CHANGELOG, 'r')
line = f.readline()
f.close()
pos = line.find('(')
posf = line.find(')', pos)
VERSION = line[pos + 1:posf].strip()
if not is_package():
    VERSION = VERSION + '-src'

####
try:
    current_locale, encoding = locale.getdefaultlocale()
    language = gettext.translation(APP, LANGDIR, [current_locale])
    language.install()
    print(language)
    if sys.version_info[0] == 3:
        _ = language.gettext
    else:
        _ = language.ugettext
except Exception as e:
    print(e)
    _ = str
APPNAME = _(APPNAME)


src/distro.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
#
# CPU-G is a program that displays information about your CPU,
# RAM, Motherboard and some general information about your System.
#
# Copyright © 2009  Fotis Tsamis <ftsamis at gmail dot com>.
# Copyright © 2016-2019  Lorenzo Carbonell (aka atareao)
# <lorenzo.carbonell.cerezo at gmail dot com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import csv

RELEASE_DATA = {}

def get_distro():
    with open("/etc/os-release") as f:
        reader = csv.reader(f, delimiter="=")
        for row in reader:
            if row:
                RELEASE_DATA[row[0]] = row[1]

    if RELEASE_DATA["ID"] in ["debian", "raspbian"]:
        with open("/etc/debian_version") as f:
            DEBIAN_VERSION = f.readline().strip()
        major_version = DEBIAN_VERSION.split(".")[0]
        version_split = RELEASE_DATA["VERSION"].split(" ", maxsplit=1)
        if version_split[0] == major_version:
            # Just major version shown, replace it with the full version
            RELEASE_DATA["VERSION"] = " ".join([DEBIAN_VERSION] + version_split[1:])
    return "{} {}".format(RELEASE_DATA["NAME"], RELEASE_DATA["VERSION"])


if __name__ == '__main__':
    print(get_distro())


src/samples.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
#
# CPU-G is a program that displays information about your CPU,
# RAM, Motherboard and some general information about your System.
#
# Copyright © 2009  Fotis Tsamis <ftsamis at gmail dot com>.
# Copyright © 2016-2019  Lorenzo Carbonell (aka atareao)
# <lorenzo.carbonell.cerezo at gmail dot com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import subprocess
import shlex
import re


def execute(command):
    command_with_args = shlex.split(command)
    execution = subprocess.Popen(command_with_args, stdout=subprocess.PIPE,
                                 stderr=subprocess.PIPE)
    stdout, stderr = execution.communicate()
    return stdout.decode()

if __name__ == '__main__':
    result = execute('/usr/bin/lspci')
    print(result)
    print(re.findall(result, '.*VGA(.*)', re.M))
    m = re.search('VGA(.*)', result)
    print(len(m.groups()))
    print(m.group(1))


src/monitor_battery.py
#!/usr/bin/env python3

import subprocess
import time
import re
import logging

if __name__ == '__main__':
    reader = open('/var/tmp/monitor_battery.log', 'r')
    data = list(reader)
    reader.close()
    if len(data) >= 600:
        writer = open('/var/tmp/monitor_battery.log', 'w')
        for line in data[300:]:
            writer.write(line)
        writer.close()

    command = ['upower', '-i', '/org/freedesktop/UPower/devices/battery_BAT0']
    ans = subprocess.check_output(command).decode()
    p = re.compile('percentage: *\d\d%')
    ans = p.findall(ans)[0][11:].strip()

    logger = logging.getLogger('monitor_battery')
    hdlr = logging.FileHandler('/var/tmp/monitor_battery.log')
    formatter = logging.Formatter('%(message)s')
    hdlr.setFormatter(formatter)
    logger.addHandler(hdlr)
    logger.setLevel(logging.INFO)
    logger.info('%s;%s' % (time.time(), ans))


src/investigator.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
#
# CPU-G is a program that displays information about your CPU,
# RAM, Motherboard and some general information about your System.
#
# Copyright © 2009  Fotis Tsamis <ftsamis at gmail dot com>.
# Copyright © 2016-2019  Lorenzo Carbonell (aka atareao)
# <lorenzo.carbonell.cerezo at gmail dot com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gi
try:
    gi.require_version('Gdk', '3.0')
except Exception as e:
    print(e)
    exit(-1)
from gi.repository import Gdk
import os
import re
import sys
import platform
import subprocess
import shlex
import comun
import psutil
from distro import get_distro
from comun import _

BATTERY_DIR = '/sys/class/power_supply/BAT0'


class Investigator():
    def readfile(self, filename):
        try:
            f = open(filename, 'r')
            data = f.read()
            f.close()
            if data.endswith('\n'):
                data = data[:-1]
            return data
        except Exception as e:
            print(e)
            return _('N/A')

    def get_window_manager(self):
        try:
            ans = self.execute('/usr/bin/wmctrl -m')
            res = ans[6:ans.find('\n')]
        except Exception as e:
            res = _('N/A')
        return res

    def resolution(self):
        s = Gdk.Screen.get_default()
        return '%dx%d' % (s.get_width(), s.get_height())

    def desktop_environment(self):
        # From http://stackoverflow.com/questions/2035657/\
        # what-is-my-current-desktop-environment
        # and http://ubuntuforums.org/showthread.php?t=652320
        # and http://ubuntuforums.org/showthread.php?t=652320
        # and http://ubuntuforums.org/showthread.php?t=1139057
        if sys.platform in ["win32", "cygwin"]:
            return "windows"
        elif sys.platform == "darwin":
            return "mac"
        else:  # Most likely either a POSIX system or something not much common
            desktop_session = os.environ.get("DESKTOP_SESSION")
            # easier to match if we doesn't have  to deal with caracter cases
            if desktop_session is not None:
                desktop_session = desktop_session.lower()
                if desktop_session in ["gnome", "unity", "cinnamon", "mate",
                                       "xfce4", "lxde", "fluxbox", "blackbox",
                                       "openbox", "icewm", "jwm",
                                       "afterstep", "trinity", "kde",
                                       "lightdm-xsession"]:
                    return desktop_session
                # ## Special cases ##
                # Canonical sets $DESKTOP_SESSION to Lubuntu rather than
                # LXDE if using LXDE.
                # There is no guarantee that they will not do the same with
                # the other desktop environments.
                elif "xfce" in desktop_session or\
                        desktop_session.startswith("xubuntu"):
                    return "xfce4"
                elif desktop_session.startswith("ubuntu"):
                    return "unity"
                elif desktop_session.startswith("lubuntu"):
                    return "lxde"
                elif desktop_session.startswith("kubuntu"):
                    return "kde"
                elif desktop_session.startswith("razor"):  # e.g. razorkwin
                    return "razor-qt"
                elif desktop_session.startswith("wmaker"):  # eg. wmaker-common
                    return "windowmaker"
            if os.environ.get('KDE_FULL_SESSION') == 'true':
                return "kde"
            elif os.environ.get('GNOME_DESKTOP_SESSION_ID'):
                if "deprecated" not in os.environ.get(
                        'GNOME_DESKTOP_SESSION_ID'):
                    return "gnome2"
            # From http://ubuntuforums.org/showthread.php?t=652320
            elif self.is_running("xfce-mcs-manage"):
                return "xfce4"
            elif self.is_running("ksmserver"):
                return "kde"
        return "unknown"

    def convert2int(value):
        try:
            return int(value)
        except Exception as e:
            return value

    def is_running(self, process):
        # From http://www.bloggerpolis.com/2011/05/\
        # how-to-check-if-a-process-is-running-using-python/
        # and http://richarddingwall.name/2009/06/18/\
        # windows-equivalents-of-ps-and-kill-commands/
        try:  # Linux/Unix
            s = subprocess.Popen(["/bin/ps", "axw"], stdout=subprocess.PIPE,
                                 universal_newlines=True)
        except Exception as e:  # Windows
            print(e)
            return False
        for x in s.stdout:
            if re.search(process, x.decode('utf-8')):
                return True
        return False

    def logo(self, core):
        model = self.cpuinfo("model", core)
        vendor = self.cpuinfo("vendor", core)
        if vendor == 'AMD':
            label = 'amd.png'
        elif vendor == 'Intel':
            label = 'intel.png'
        # AMDs
        if re.match("AMD Athlon\(tm\) 64 X2.*", model)\
                or re.match("AMD Athlon\(tm\) X2.*", model):
            label = 'AMD-AthlonX2.png'
        elif re.match("AMD Sempron\(tm\).*", model):
            label = 'AMD-Sempron.png'
        elif re.match("Mobile AMD Sempron\(tm\).*", model):
            label = 'AMD-Sempron-Mobile.png'
        elif re.match("Dual-Core AMD Opteron\(tm\).*", model)\
                or re.match("AMD Opteron\(tm\).*", model):
            label = 'AMD-Opteron.png'
        elif re.match("AMD Athlon\(tm\) XP.*", model):
            label = 'AMD-AthlonXP.png'
        elif re.match("AMD Athlon\(tm\) 64 Processor.*", model):
            label = 'AMD-Athlon64.png'
        elif re.match("AMD Phenom\(tm\).*", model):
            label = 'AMD-Phenom.png'
        # Intels
        elif re.match("Intel\(R\) Core\(TM\)2 Duo.*", model):
            label = 'Intel-Core2Duo.png'
        elif re.match("Intel\(R\) Core\(TM\)2 Quad.*", model):
            label = 'Intel-Core2Quad.png'
        elif re.match("Intel\(R\) Core\(TM\)2 CPU.*", model):
            label = 'Intel-Core2Quad.png'
        elif re.match("Intel\(R\) Atom\(TM\) CPU.*", model):
            label = 'Intel-Atom.png'
        elif re.match("Intel\(R\) Core\(TM\)2 Extreme CPU.*", model):
            label = 'Intel-Core2Extreme.png'
        elif re.match("Intel\(R\) Xeon\(TM\).*", model):
            label = 'Intel-Xeon.png'
        elif re.match(".*Pentium II.*", model):
            label = 'Intel-Pentium2.png'
        elif re.match("Intel\(R\) Pentium\(R\) Dual CPU.*", model):
            label = 'Intel-PentiumDual.png'
        return label

    def hostname(self):
        ans = self.readfile('/etc/hostname')
        if ans.endswith('\n'):
            ans = ans[:-1]
        return ans

    def battery_info(self, data='manufacturer'):
        info = 0
        if data == 'manufacturer':
            info = self.readfile(os.path.join(BATTERY_DIR, 'manufacturer'))
        elif data == 'model_name':
            info = self.readfile(os.path.join(BATTERY_DIR, 'model_name'))
        elif data == 'serial_number':
            info = self.readfile(os.path.join(BATTERY_DIR, 'serial_number'))
        elif data == 'technology':
            info = self.readfile(os.path.join(BATTERY_DIR, 'technology'))
        elif data == 'status':
            info = self.readfile(os.path.join(BATTERY_DIR, 'status'))
        elif data == 'capacity':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'capacity')))
        elif data == 'capacity-level':
            info = self.readfile(os.path.join(BATTERY_DIR, 'capacity_level'))
        elif data == 'voltage-now':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'voltage_now')))
        elif data == 'voltage-min-design':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'voltage_min_design')))
        elif data == 'charge-now':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'charge_now')))
        elif data == 'current-now':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'current_now')))
        elif data == 'charge-full':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'charge_full')))
        elif data == 'charge-full-design':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'charge_full_design')))
        elif data == 'cycle-count':
            info = Investigator.convert2int(
                self.readfile(os.path.join(BATTERY_DIR, 'cycle_count')))
        elif data == 'present':
            info = (self.readfile(os.path.join(BATTERY_DIR,
                                               'present')) == 1)
        elif data == 'exists':
            info = os.path.exists(BATTERY_DIR)
        return info

    def cpuinfo(self, var, core=0):
        info = self.readfile("/proc/cpuinfo")

        if var == 'vendor':
            vendor = re.findall("vendor_id\s*:\s*(.*)", info)
            if vendor[core] == 'AuthenticAMD':
                vendor[core] = 'AMD'
            elif vendor[core] == 'GenuineIntel':
                vendor[core] = 'Intel'
            return vendor[core]
        elif var == 'corespeed':
            return re.findall("cpu MHz\s*:\s*(.*)", info)[core] + ' MHz'
        elif var == 'model':
            return re.findall("model name\s*:\s*(.*)", info)[core]
        elif var == 'cache':
            return re.findall("cache size\s*:\s*(.*)", info)[core]
        elif var == 'modelnumber':
            return re.findall("model\s*:\s*(.*)", info)[core]
        elif var == 'family':
            return re.findall("cpu family\s*:\s*(.*)", info)[core]
        elif var == 'stepping':
            return re.findall("stepping\s*:\s*(.*)", info)[core]
        elif var == 'coresnum':
            return str(len(re.findall("processor\s*:\s*(.*)", info)))
        elif var == 'flags':
            return re.findall("flags\s*:\s*(.*)", info)[core]
        elif var == 'bogomips':
            return re.findall("bogomips\s*:\s*(.*)", info)[core]
        elif var == 'width':
            if re.findall(' lm(?![-a-zA-Z0-9_])',
                          re.findall("flags\s*:(.*)", info)[core]):
                width = '64-bit'
            else:
                width = '32-bit'
            return width

    def sysdevcpu(self, core, level, kind):
        if not os.path.exists('/sys/devices/system/cpu'):
            return _('N/A')
        coresinsysdev = str(
            len(re.findall("'cpu[0-9]+'",
                           str(os.listdir("/sys/devices/system/cpu/")))))
        if coresinsysdev == self.cpuinfo('coresnum'):
            cores_matching = True
        else:
            print("Error: Cannot decide if the cores are %s or %s.\n" +
                  "Using the lowest value as the real cores number." %
                  (self.cpuinfo('coresnum'), coresinsysdev))
        path = '/sys/devices/system/cpu/cpu%i/cache/' % (core)
        indexes = len(
            re.findall("'index[0-9]*'", str(os.listdir(path))))
        for index in range(indexes):
            levelpath = path + 'index%i/level' % (index)
            typepath = path + 'index%i/type' % (index)
            size = path + 'index%i/size' % (index)
            # os.chdir(newpath)
            if self.readfile(levelpath).strip() == str(level) and\
                    self.readfile(typepath).strip() == kind:
                return self.readfile(size).strip()
        return _('N/A')

    def distro(self):
        return get_distro()
        try:
            values = platform.linux_distribution()
        except AttributeError:
            values = platform.dist()
        if len(values) == 3:
            return "%s %s %s" % (values[0], values[1], values[2])
        return self.readfile('/etc/issue').strip()

    def get_distro_logo(self):
        distro = self.distro().split()[0].lower() + '.png'
        logo = os.path.join(comun.DISTROSDIR, distro)
        if os.path.exists(logo):
            return logo
        return None

    def gccver(self):
        gcc_version = self.execute('/usr/bin/gcc -dumpversion').strip()
        if gcc_version == '':
            gcc_version = _('N/A')
        return gcc_version

    def xver(self):
        try:
            xver_version = self.execute('/usr/bin/apt-cache show xorg')
        except FileNotFoundError:
            xver_version = ""
        ans = re.findall("Version: 1:(.*)\+", xver_version)
        if ans:
            return ans[0]
        return _('N/A')

    def raminfo(self):
        mem = psutil.virtual_memory()
        values = {'total': mem.total/1024/1024,
                  'available': mem.available/1024/1024,
                  'percent': mem.percent,
                  'used': mem.used/1024/1024,
                  'free': mem.free/1024/1024,
                  'active': mem.active/1024/1024,
                  'inactive': mem.inactive/1024/1024,
                  'buffers': mem.buffers/1024/1024,
                  'cached': mem.cached/1024/1024,
                  }
        return values

    def swapinfo(self):
        mem = psutil.swap_memory()
        values = {'total': mem.total/1024/1024,
                  'used': mem.used/1024/1024,
                  'free': mem.free/1024/1024,
                  'sin': mem.sin/1024/1024,
                  'sout': mem.sout/1024/1024,
                  }
        return values

    def disksinfo(self):
        values = []
        disk_partitions = psutil.disk_partitions(all=False)
        for partition in disk_partitions:
            usage = psutil.disk_usage(partition.mountpoint)
            device = {'device': partition.device,
                      'mountpoint': partition.mountpoint,
                      'fstype': partition.fstype,
                      'opts': partition.opts,
                      'total': usage.total,
                      'used': usage.used,
                      'free': usage.free,
                      'percent': usage.percent
                      }
            values.append(device)
        values = sorted(values, key=lambda device: device['device'])
        return values

    def mobo(self, var):
        # var can be: board_vendor, board_name, bios_vendor,
        # bios_version, bios_date, or chassis_type
        return self.readfile('/sys/devices/virtual/dmi/id/' + var).strip()

    def uptime(self):
        total = int(self.readfile('/proc/uptime').split('.')[0])
        days = int(total / 86400)
        hours = int((total / 3600) - (days * 24))
        minutes = int((total / 60) - ((days * 1440) + (hours * 60)))
        return "%i days, %i hours, %i minutes" % (days, hours, minutes)

    def get_graphic_card_logo(self):
        card_logo = self.execute('/usr/bin/lspci', 'VGA(.*)')
        if card_logo is not None:
            # Intel
            if re.findall("Intel\s*", card_logo):
                label = 'intel.png'
            # ATI
            # ATI Technologies replace to ATI. See bug
            # https://bugs.launchpad.net/cpug/+bug/959115
            elif re.findall("ATI\s*", card_logo):
                label = 'ati.png'
            # nVidia
            # elif re.findall("nVidia\s*", card_logo):
            elif re.findall("nVidia\s*", card_logo, re.I):
                label = 'nvidia.png'
            else:
                label = 'unknown.png'
        else:
            label = 'unknown.png'
        return os.path.join(comun.GRAPHICCARDDIR, label)

    # Graphic tab
    def open_gl(self, var):
        open_gl_ = self.execute('/usr/bin/glxinfo')
        vga = self.execute('/usr/bin/lspci', 'VGA[^:]*:(.*)')
        if var == 'vendor':
            if open_gl_ != '':
                return re.findall("OpenGL vendor string: (.*)", open_gl_)[0]
            return _('N/A')
        elif var == 'renderer':
            if open_gl_ != '':
                return re.findall("OpenGL renderer string: (.*)", open_gl_)[0]
            return _('N/A')
        elif var == 'version':
            if open_gl_ != '':
                return re.findall("OpenGL version string: (.*)", open_gl_)[0]
            return _('N/A')
        elif var == 'VGA':
            if vga is not None:
                return vga
            return _('N/A')
    # End Graphic Tab

    def execute(self, command, parser=None):
        command_with_args = shlex.split(command)
        execution = subprocess.Popen(command_with_args, stdout=subprocess.PIPE,
                                     stderr=subprocess.PIPE)
        stdout, stderr = execution.communicate()
        if parser is None:
            result = stdout.decode()
        else:
            matches = re.search(parser, stdout.decode())
            if matches.groups():
                result = matches.group(1)
            else:
                result = None
        return result

if __name__ == '__main__':
    print(psutil.cpu_times())
    print(psutil.cpu_count())
    print(psutil.virtual_memory())
    print(psutil.swap_memory())
    print(psutil.disk_partitions())
    print(psutil.disk_usage('/'))
    # exit(0)
    inv = Investigator()
    print(inv.cpuinfo)
    # board_vendor, board_name, bios_vendor, bios_version, bios_date,\
    # chassis_type
    print(inv.cpuinfo('coresnum'))
    for i in range(0, int(inv.cpuinfo('coresnum'))):
        print(inv.cpuinfo('vendor', i))
        print(inv.cpuinfo('corespeed', i))
        print(inv.cpuinfo('model', i))
        print(inv.cpuinfo('cache', i))
        print(inv.cpuinfo('modelnumber', i))
        print(inv.cpuinfo('family', i))
        print(inv.cpuinfo('stepping', i))
        print(inv.cpuinfo('flags', i))
        print(inv.cpuinfo('bogomips', i))
        print(inv.cpuinfo('width', i))
        print(inv.logo(i))
        print(inv.sysdevcpu(i, 1, 'Data'))
        print(inv.sysdevcpu(i, 1, 'Instruction'))
        print(inv.sysdevcpu(i, 2, 'Unified'))
        print(inv.sysdevcpu(i, 3, 'Unified'))
    print(inv.mobo('board_vendor'))
    print(inv.mobo('board_name'))
    print(inv.mobo('bios_vendor'))
    print(inv.mobo('bios_version'))
    print(inv.mobo('chassis_type'))
    print(inv.raminfo())
    print(inv.distro())
    print(inv.xver())
    print(inv.gccver())
    print(inv.disksinfo())
    print(inv.get_window_manager())

    exit(0)
